### ISSUE TITLE
Use downward API with volume mount to monitor pod labels/annotations

### ISSUE THREAD

**Issue by rshriram** [istio/istio#11893]
The labels and annotations in a pod are currently detected statically during injection or via helm spec and encoded into  ISTIO_META_INSTANCE_LABELS/ANNOTATIONS. These metadata are currently not the authoritative source since pod labels and annotations could change dynamically. So we are forced to monitor k8s api server for label updates as well.

K8S allows mounting all the pod labels/annotations as a "file" using the downward API. I verified that this works for deployments as well. For example,

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: httpbin
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: httpbin
        version: v1
      annotations:
        sidecar.istio.io/inject: "false"
    spec:
      containers:
      - image: docker.io/kennethreitz/httpbin
        imagePullPolicy: IfNotPresent
        name: httpbin
        ports:
        - containerPort: 80
        volumeMounts:
        - name: podinfo
          mountPath: /tmp/etc
          readOnly: true
      volumes:
      - name: podinfo
        downwardAPI:
          defaultMode: 444
          items:
          - fieldRef:
              fieldPath: metadata.labels
            path: pod_labels
          - fieldRef:
              fieldPath: metadata.annotations
            path: pod_annotations
```

Rather than detecting the labels at injection time, we could monitor the file path /tmp/etc/pod_labels, obtain the list of labels and add them to the metadata before starting up Envoy. The mounted files change when you add/remove labels to a pod (verified). Curiously, when you add/remove labels from a deployment, the entire deployment is restarted [which is good for us].

Bottom line is that we now have a reliable way of obtaining pod labels from the pod/proxy metadata while being able to monitor for changes. When the labels change, the proxy needs to be restarted with the new set of labels encoded in the metadata. 

It is not clear where this will be useful as we still rely on the API server to get list of all service instances with their labels, to form the subsets for destination rules.

**Comment by rshriram** [comment]
cc @hzxuzhonghu @costinm @mandarjog

**Comment by douglas-reid** [comment]
Does this provide a way to have, say, authoritative source labels injected on requests?  If so, at a minimum, it gives us a way to provide richer telemetry in multi-cluster setups where the local mixer doesn't have remote API server access.

**Comment by hzxuzhonghu** [comment]
>Curiously, when you add/remove labels from a deployment, the entire deployment is restarted [which is good for us].

This is the intended behavior for k8s.  K8s discourages update pod directly which belong ro a Deployment or RS.

But overall `downward API with volume mount` can help proxy update labels/annotations. Which currently dont support I think.

**Comment by hzxuzhonghu** [comment]
Agree, we should watch the mounted files and restart envoy when changes.

**Comment by yangminzhu** [comment]
/cc @yangminzhu @diemtvu 

> Does this provide a way to have, say, authoritative source labels injected on requests?

@douglas-reid I think here we're only changing where to get the labels. Before we get it in the sidecar-injector and now we're trying to get it directly from the k8s api server via the mounted file. There is no difference after this step. So if we want to put the label in the request, we still need to do some other work.

We will need this for the migration of the auth v2 policy from service based to label based selector. Currently this is not the top priority for the auth v2 policy, we're still implementing the basic support, but we need to solve this eventually. If no one is going to work on this, I can take it for Istio 1.2 or 1.3.

**Comment by douglas-reid** [comment]
@yangminzhu well, right now, labels are coming from Mixer's k8s adapter in most policy/telemetry scenarios. and it is coming from the api server (not from the sidecar injector). if we can reliably get authoritative pod labels at the proxy, the mixer client can put that info into the headers for http/grpc.  this also has potential implications for simplifying mixer v2 approaches.

/cc @mandarjog
